pkg/controller/volume/attachdetach/util/util_test.go

```diff
@@ -0,0 +1,316 @@
+/*
+Copyright 2017 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package util
+
+import (
+	"os"
+	"reflect"
+	"testing"
+	"time"
+
+	v1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	kubetypes "k8s.io/apimachinery/pkg/types"
+	"k8s.io/client-go/informers"
+	csitrans "k8s.io/csi-translation-lib"
+	fakeframework "k8s.io/kubernetes/pkg/scheduler/framework/fake"
+	"k8s.io/kubernetes/pkg/volume/csimigration"
+	"k8s.io/kubernetes/pkg/volume/fc"
+
+	utilfeature "k8s.io/apiserver/pkg/util/feature"
+	"k8s.io/client-go/kubernetes/fake"
+	utiltesting "k8s.io/client-go/util/testing"
+	"k8s.io/kubernetes/pkg/volume"
+	volumetest "k8s.io/kubernetes/pkg/volume/testing"
+)
+
+const (
+	testHostName = "test-hostname"
+	socketPath   = "/var/run/kmsplugin"
+	volumeName   = "volume-name"
+	testNodeName = "test-node-name"
+)
+
+func Test_CreateVolumeSpec(t *testing.T) {
+	var (
+		dirOrCreate = v1.HostPathType(v1.HostPathDirectoryOrCreate)
+		nodeName    = kubetypes.NodeName(testNodeName)
+		hostPath    = &v1.HostPathVolumeSource{
+			Path: socketPath,
+			Type: &dirOrCreate,
+		}
+		objectReference = v1.ObjectReference{Namespace: "claimrefns", Name: "test-pvc"}
+
+		volumeExampe = v1.Volume{
+			Name: volumeName,
+			VolumeSource: v1.VolumeSource{
+				HostPath: hostPath,
+			},
+		}
+		volumeExampeWithCSI = v1.Volume{
+			Name: volumeName,
+			VolumeSource: v1.VolumeSource{
+				PersistentVolumeClaim: &v1.PersistentVolumeClaimVolumeSource{
+					ClaimName: "test-pvc",
+					// TODO: Use source definition to set this value when we have one.
+					ReadOnly: false,
+				},
+			},
+		}
+		persistentVolumeExampeWithCSI = v1.PersistentVolume{
+			ObjectMeta: metav1.ObjectMeta{
+				Name: volumeName,
+			},
+			Spec: v1.PersistentVolumeSpec{
+				PersistentVolumeSource: v1.PersistentVolumeSource{
+					CSI: &v1.CSIPersistentVolumeSource{
+						Driver:       "pd.csi.storage.gke.io",
+						VolumeHandle: "projects/UNSPECIFIED/zones/UNSPECIFIED/disks/test-disk",
+						FSType:       "ext4",
+						ReadOnly:     false,
+						VolumeAttributes: map[string]string{"partition": ""},
+					},
+				},
+				ClaimRef: &objectReference,
+			},
+		}
+	)
+
+	tmpDir, err := utiltesting.MkTmpdir("csi-test")
+	if err != nil {
+		t.Fatalf("can't make a temp dir: %v", err)
+	}
+	defer os.RemoveAll(tmpDir)
+
+	csiTranslator := csitrans.New()
```

Comment:
All the initialization should be done per test case because they could be run in parallel. You could put these into a "setup()" function for better readability.